### PR TITLE
Avoid holding onto bulk items until all completed

### PR DESCRIPTION
**Pull request by ywelsch**
Bulk requests currently keep a reference to all bulk item requests until every one of them has completed. There is no need to do so, however, and, in case of large bulks, can mean unnecessary holding onto memory that might be better used elsewhere. More so as different shard-level bulks can complete at different speeds, and one slow shard-level request should not require holding onto every other shard-level request.

### CODE REVIEW

**Review by original-brownbear — APPROVED** [review]
LGTM, nice find :)

**Comment by ywelsch** [comment]
I had to implement this differently as mutating the request object had undesired effects with the RetryHandler class that's used by Update/Delete-by-Query, which makes uses of the request for retries.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 1]
```diff
@@ -388,7 +388,7 @@ private long buildTookInMillis(long startTimeNanos) {
      * */
     private final class BulkOperation extends ActionRunnable<BulkResponse> {
         private final Task task;
-        private final BulkRequest bulkRequest;
+        private volatile BulkRequest bulkRequest; // set to null once all requests are sent out
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: NIT: no need for this to be `volatile` I think, you're only ever touching it on the thread that nulls it out?

**Review by original-brownbear — APPROVED** [review]
LGTM as well, just one suggestion/question :)

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Pull request merged**